File: apps/backend/app/core/retry_policy.py
import math
import logging
from typing import Dict, Any, Optional

logger = logging.getLogger("retry-policy")

class RetryPolicy:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(RetryPolicy, cls).__new__(cls)
            cls._instance.attempts: Dict[str, int] = {}
        return cls._instance

    def evaluate(
        self,
        run_id: str,
        failure_type: str,
        current_agent: Optional[str] = None,
        max_retries: int = 3
    ) -> Dict[str, Any]:
        key = f"{run_id}:{failure_type}"
        count = self.attempts.get(key, 0) + 1
        self.attempts[key] = count

        logger.info(f"Evaluating failure retry policy for run={run_id}, attempts={count}")

        if count > max_retries:
            return {
                "action": "terminate",
                "delay_ms": 0,
                "reason": f"Exceeded maximum retries ({max_retries}) for failure type '{failure_type}'"
            }

        if failure_type == "selector" and current_agent == "browser-execution":
            return {
                "action": "switch_agent",
                "delay_ms": 500,
                "suggested_agent": "failure-analysis",
                "reason": "Locator failure detected. Switching to Failure Diagnostics Agent."
            }

        delay_ms = int(math.pow(2, count) * 1000)
        return {
            "action": "exponential_backoff",
            "delay_ms": delay_ms,
            "reason": f"Retrying after {delay_ms}ms backoff (Attempt {count}/{max_retries})"
        }

retry_policy = RetryPolicy()

File: packages/agents/src/policy/retry-policy.ts
import type { RetryDecision, AgentType } from "@testpilot/types";
import { createLogger } from "@testpilot/shared";

const logger = createLogger("retry-policy");

/**
 * Retry Policy Engine: Evaluates failure events and decides retry actions,
 * exponential backoffs, agent switches, or execution terminations.
 */
export class RetryPolicy {
  private static instance: RetryPolicy;
  private attemptCounts: Map<string, number> = new Map();

  private constructor() {}

  public static getInstance(): RetryPolicy {
    if (!RetryPolicy.instance) {
      RetryPolicy.instance = new RetryPolicy();
    }
    return RetryPolicy.instance;
  }

  /** Evaluate failure context and determine appropriate recovery policy */
  public evaluate(
    runId: string,
    failureType: string,
    currentAgent?: AgentType,
    maxRetries = 3
  ): RetryDecision {
    const key = `${runId}:${failureType}`;
    const attempts = (this.attemptCounts.get(key) || 0) + 1;
    this.attemptCounts.set(key, attempts);

    logger.info({ runId, failureType, attempts, maxRetries }, "Evaluating failure retry policy");

    if (attempts > maxRetries) {
      return {
        action: "terminate",
        delayMs: 0,
        reason: `Maximum retry limit (${maxRetries}) exceeded for failure type '${failureType}'`,
      };
    }

    if (failureType === "selector" && currentAgent === "browser-execution") {
      return {
        action: "switch_agent",
        delayMs: 500,
        suggestedAgent: "failure-analysis",
        reason: "Locator failure detected. Switching from Browser Execution to Failure Diagnostics Agent.",
      };
    }

    // Default exponential backoff
    const delayMs = Math.pow(2, attempts) * 1000;
    return {
      action: "exponential_backoff",
      delayMs,
      reason: `Retrying step after ${delayMs}ms exponential backoff delay (Attempt ${attempts}/${maxRetries})`,
    };
  }

  /** Reset retry counts for a run */
  public reset(runId: string): void {
    for (const key of this.attemptCounts.keys()) {
      if (key.startsWith(`${runId}:`)) {
        this.attemptCounts.delete(key);
      }
    }
  }
}

export const retryPolicy = RetryPolicy.getInstance();
